fix(tags): Refactor to fetch tags with respect to the global selection header

dashed opened the pull request:
**WORK IN PROGRESS**

## TODO

- [ ] store invalidation after N minutes?
- [x] add test for `extractDatetimeSelectionParameters`
- [x] ~bug: global selection store does not get updated whenever relevant query strings changes~ (EDIT: not expected to be addressed in this PR)
- [x] clean up: delete `fetchOrganizationTags`

dashed committed:
s

dashed committed:
loadOrganizationTags

dashed committed:
src/sentry/static/sentry/app/views/issueList/overview.jsx

dashed committed:
MAX_TAGS = 1000

dashed committed:
return promise for now

dashed committed:
src/sentry/static/sentry/app/views/events/searchBar.jsx

dashed committed:
things

dashed committed:
rm console.log

dashed committed:
rename

dashed committed:
parse utc only when it's defined

dashed committed:
better than nothing

dashed committed:
extractDatetimeSelectionParameters

dashed committed:
add more edge test cases for getParams

dashed committed:
hmm

dashed committed:
alternative

dashed commented on src/sentry/static/sentry/app/views/eventsV2/table/index.tsx:
```diff
@@ -162,7 +172,7 @@ class Table extends React.PureComponent<TableProps, TableState> {
 }
 
-export default withApi<TableProps>(Table);
+export default withApi(Table);
```
@dashed revert

markstory commented on src/sentry/static/sentry/app/actionCreators/tags.jsx:
```diff
@@ -60,6 +61,34 @@ function tagFetchSuccess(tags) {
   TagActions.loadTagsSuccess(trimmedTags);
 }
 
+export function loadOrganizationTags(api, orgId, selection) {
```
The `fetchOrganizationTags` function in this file has a TODO to add more global selection support. This might be a good time to do that over making new functions.

dashed replied:
This function was meant to replace `fetchOrganizationTags`. At the moment, I'm facing issues with stale values (i.e. out of sync from the `location` object) from the global selection store being improperly propagated to React components. This incurs an additional unnecessary API call 😞 . 

I'm going to try a different approach of: 

1. Have `tagStore` listen to `globalSelectionStore` directly (yay flux)
2. Create `withOrganizationTags` HoC that re-renders whenever `tagStore` is updated
3. Refactor out `fetchOrganizationTags`/`loadOrganizationTags` from `sentry`; with `withOrganizationTags` as the replacement.

dashed commented:
@markstory to take over this PR.

markstory committed:
Merge branch 'master' into refactor-fetch-tags

markstory committed:
Refactor organization tag loading into an HOC

Using an HOC allows us to leverage the existing reflux stores more
effectively and better manage how often an organization's tag list is
reloaded.

markstory commented:
@dashed I think I got the various problems in your todo list taken care of.

dashed commented:
Sweet.

markstory committed:
Don't do API requests in the HOC

The HOC can be used multiple times in a render tree and we don't want
multiple requests sent. Instead update all the top level components to
initiate requests to populate the store and have withTags() listen to
store changes. This pattern is used quite extensively by our other HOC
components.

markstory commented:
cc @billyvg I had to change incident rules forms as they were using the events/searchBar component. I tested locally and ensured that tag keys showed up in the autocomplete.

markstory committed:
Merge branch 'master' into refactor-fetch-tags

markstory committed:
Merge branch 'master' into refactor-fetch-tags

wmak commented on src/sentry/static/sentry/app/actionCreators/tags.jsx:
```diff
@@ -60,6 +61,43 @@ function tagFetchSuccess(tags) {
+  TagStore.reset();
+
+  const url = `/organizations/${orgId}/tags/`;
+  const query = selection.datetime ? {...getParams(selection.datetime)} : {};
```
could we add `use_cache: 1` here?

markstory committed:
Merge branch 'master' into refactor-fetch-tags

markstory committed:
Fix prettier error.

markstory committed:
Add missing cache enable flag.

markstory committed:
Merge branch 'master' into refactor-fetch-tags
